meta-mender-core/scripts/create-installer.sh

```diff
@@ -0,0 +1,268 @@
+    INACTIVE_PART_START=\${PART_A_START_MB}
+else
+    INACTIVE_PART_START=
+fi
+
+# Verify that the inactive partition is known
+if [ -z "\${INACTIVE_PART_START}" ]; then
+    echo "Unable to determine the partition being copied from root device \"\${ROOT_DEVICE}\""
+    exit 1
+fi
+
+#
+# Wake the eMMC. Before adding this, writing the disk label would produce the
+# following kernel errors and break the install:
+#
+# [   38.274535] mmc1: cache flush error -110
+# [   38.279509] print_req_error: I/O error, dev mmcblk1, sector 0
+#
+dd if="\${INSTALL_DEVICE}" of="/dev/null" bs=512 count=1
+
+# Copy the bootloaded and SPL, if present, embedded in the space before the
+# boot partition
+dd if="\${STORAGE_DEVICE}" \\
+  of="\${INSTALL_DEVICE}" \\
+  bs=1M \\
+  count=\${BOOT_PART_START_MB}
+
+# Write a partition table to the disk
+parted -s "\${INSTALL_DEVICE}" mklabel "\${PARTITION_TABLE_TYPE}"
+sync
+
+# Create boot partition 1
+parted -s "\${INSTALL_DEVICE}" -a min unit s mkpart primary fat32 "\${BOOT_PART_START_SECTORS}" "\${BOOT_PART_END_SECTORS}"
+sync
+parted -s "\${INSTALL_DEVICE}" set 1 boot on
+sync
+
+# Create partition A
+parted -s "\${INSTALL_DEVICE}" -a min unit s mkpart primary ext4 "\${PART_A_START_SECTORS}" "\${PART_A_END_SECTORS}"
+sync
+
+# Create partition B
+parted -s "\${INSTALL_DEVICE}" -a min unit s mkpart primary ext4 "\${PART_B_START_SECTORS}" "\${PART_B_END_SECTORS}"
+sync
+
+# Create data partition
+parted -s "\${INSTALL_DEVICE}" -a min unit s mkpart primary ext4 "\${DATA_PART_START_SECTORS}" "\${DATA_PART_END_SECTORS}"
+sync
+
+# Copy boot partition
+dd if="\${STORAGE_DEVICE}" \\
+  of="\${INSTALL_DEVICE}" \\
+  bs=1M \\
+  skip=\${BOOT_PART_START_MB} \\
+  seek=\${BOOT_PART_START_MB} \\
+  count=\${BOOT_PART_SIZE_MB}
+
+# Check boot partition
+#
+# Note the mtools 4.0.19 (in Yocto Warrior) corrupts the EFI directory when
+# generating the image. Switch to mtools 4.0.23 (in Yocto Zeus).
+#
+# We don't abort here, because fsck.fat will always return 1 given that this
+# script is run from the boot partition which will have the dirty bit set.
+#
+# Instead, we could abort if the output contains an error string. Some observed
+# ones are:
+#
+#     Bad short file name (.).
+#     File size is 0 bytes, cluster chain length is > 0 bytes.
+#     Contains a free cluster (3). Assuming EOF.
+#
+fsck.fat -a -v -l "${INSTALL_DEVICE}p1" || :
+
+# TODO: Until U-Boot can pass disk context to GRUB, we need to manually update
+# grub.cfg to pass the correct disk to the kernel
+mkdir -p "\${INSTALL_BOOT_MOUNT}"
+mount "\${INSTALL_DEVICE}p1" "\${INSTALL_BOOT_MOUNT}"
+sed -i "s#\${STORAGE_DEVICE}#\${INSTALL_DEVICE}#g" "\${INSTALL_BOOT_MOUNT}/EFI/BOOT/grub.cfg"
+sync
+umount "\${INSTALL_BOOT_MOUNT}"
+rm -rf "\${INSTALL_BOOT_MOUNT}"
+
+# Copy root partition twice
+for dest in \${PART_A_START_MB} \${PART_B_START_MB}; do
+    dd if="\${STORAGE_DEVICE}" \\
+      of="\${INSTALL_DEVICE}" \\
```

Comment:
I'll look into `mender snapshot dump`, but this call might be safe because the line below is `skip=\${INACTIVE_PART_START}` - it copies the inactive partition twice. However, it copies the boot partition above which is mounted RW, but I'm assuming we don't write to `/boot` often for safety. In that case fsck is mostly used to reset the dirty bit.

Reply:
Oh, you're right, I missed that. Still, using `snapshot dump` could be useful if you have modified the active partition to install a package or something like that. And it will make the `/boot` access safe as well.